source/ambassador_host.go

```diff
@@ -0,0 +1,265 @@
+// The IngressRoute implementation uses the spec.virtualHost.fqdn value for the hostname.
+// Use targetAnnotationKey to explicitly set Endpoint.
+type ambassadorHostSource struct {
+	dynamicKubeClient      dynamic.Interface
+	kubeClient             kubernetes.Interface
+	namespace              string
+	ambassadorHostInformer informers.GenericInformer
+}
+
+//
+// An Ambassador Host resource looks like this:
+//
+//  apiVersion: getambassador.io/v2
+//  kind: Host
+//  metadata:
+//    name: e78cb61d56b6.eu.ngrok.io
+//  spec:
+//    acmeProvider:
+//      authority: https://acme-v02.api.letsencrypt.org/directory
+//      email: alvaro.saurin@gmail.com
+//    ambassadorId:
+//    - default
+//    hostname: e78cb61d56b6.eu.ngrok.io
+//    selector:
+//      matchLabels:
+//        hostname: e78cb61d56b6.eu.ngrok.io
+//    tlsSecret:
+//      name: e78cb61d56b6.eu.ngrok.io
+//  status:
+//    state: Ready
+//    tlsCertificateSource: ACME
+//
+
+// NewAmbassadorHostSource creates a new ambassadorHostSource with the given config.
+func NewAmbassadorHostSource(
+	dynamicKubeClient dynamic.Interface,
+	kubeClient kubernetes.Interface,
+	namespace string) (Source, error) {
+	var err error
+
+	// Use shared informer to listen for add/update/delete of Host in the specified namespace.
+	// Set resync period to 0, to prevent processing when nothing has changed.
+	informerFactory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(dynamicKubeClient, 0, namespace, nil)
+	ambassadorHostInformer := informerFactory.ForResource(ambHostGVR)
+
+	// Add default resource event handlers to properly initialize informer.
+	ambassadorHostInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+			},
+		},
+	)
+
+	// TODO informer is not explicitly stopped since controller is not passing in its channel.
+	informerFactory.Start(wait.NeverStop)
+
+	// wait for the local cache to be populated.
+	err = poll(time.Second, 60*time.Second, func() (bool, error) {
+		return ambassadorHostInformer.Informer().HasSynced(), nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to sync cache: %v", err)
+	}
+
+	return &ambassadorHostSource{
+		dynamicKubeClient:      dynamicKubeClient,
+		kubeClient:             kubeClient,
+		namespace:              namespace,
+		ambassadorHostInformer: ambassadorHostInformer,
+	}, nil
+}
+
+// Endpoints returns endpoint objects for each host-target combination that should be processed.
+// Retrieves all Hosts in the source's namespace(s).
+func (sc *ambassadorHostSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	hosts, err := sc.ambassadorHostInformer.Lister().ByNamespace(sc.namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	endpoints := []*endpoint.Endpoint{}
+
+	for _, host := range hosts {
+		unstructuredHost, ok := host.(*unstructured.Unstructured)
```

Comment:
Can't this be structured?

Reply:
Ambassador's `Host`s are currently not defined in any Go library. We could define it here but I'm concerned it could break any deserialization if the `Host` structure changes. Not sure it is really worth it...

Reply:
New Go bindings have been published in [/pkg/api/getambassador.io/v2](https://github.com/datawire/ambassador/tree/master/pkg/api/getambassador.io/v2). I've just updated the code for deserializing to a `Host` resource.